File: probfilter-core/src/main/java/com/c6h5no2/probfilter/util/SimpleLCG.java
package com.c6h5no2.probfilter.util;

import java.io.Serial;


/**
 * A linear congruential generator.
 */
public final class SimpleLCG implements RandomIntGenerator {
    @Serial
    private static final long serialVersionUID = 1L;

    private int state;

    public SimpleLCG(int seed) {
        this.state = ~seed;
    }

    @Override
    public int nextInt() {
        state = (state * 1103515245 + 12345) & Integer.MAX_VALUE;
        return state;
    }

    @Override
    public SimpleLCG copy() {
        var rng = new SimpleLCG(0);
        rng.state = this.state;
        return rng;
    }
}

File: probfilter-core/src/main/scala/com/c6h5no2/probfilter/util/SimpleLCG.scala
package com.c6h5no2.probfilter.util


/**
 * A linear congruential generator.
 */
@SerialVersionUID(1L)
class SimpleLCG(seed: Int) extends RandomIntGenerator {
  private[this] val state = ~seed

  override def nextState(): RandomIntGenerator = {
    val next = (state * 1103515245 + 12345) & Int.MaxValue
    new SimpleLCG(~next) // ~~=0
  }

  override def getInt: Int = state
}
